tmpl/blog/mirage-3.0-errors.md

```diff
@@ -0,0 +1,192 @@
+exhaustiveness checks mean we can force callers to consider errors
+where necessary.
+
+There are three main aspects to consider:
+
+- Providing diagnostic information to humans.
+- Allowing programs to detect and handle certain errors specially.
+- Indicating failure in order to abort or roll back the current operation.
+
+Note that the last two points are very different. Exceptions can be
+thrown at any point in an OCaml program and robust code must be
+prepared to handle this. In particular, this (fictional) code is
+*wrong*:
+
+```
+  let gntref = Gntshr.get () in
+  match_lwt f gntref with
+  | `Ok () -> Gntshr.put gntref; return (`Ok ())
+  | `Error _ as e -> Gntshr.put gntref; return e
+```
+
+If `f` throws an exception, the grant ref will be leaked.
+The correct pattern is:
+
+```
+  Gntshr.with_ref f
+```
+
+This will release the resource whether `f` returns success, returns an
+error code or raises an exception.
+
+### Errors vs. Exceptions
+
+Real World OCaml's [Chapter 7. Error Handling][rwo-errors] provides an
+excellent overview of the options for handling errors in OCaml. It
+finishes with this good advice:
+
+> To be clear, it doesn't make sense to avoid exceptions entirely. The
+> maxim of "use exceptions for exceptional conditions" applies. If an
+> error occurs sufficiently rarely, then throwing an exception is often
+> the right behavior.
+>
+> Also, for errors that are omnipresent, error-aware return types may
+> be overkill. A good example is out-of-memory errors, which can occur
+> anywhere, and so you'd need to use error-aware return types
+> everywhere to capture those.  **Having every operation marked as one
+> that might fail is no more explicit than having none of them
+> marked.**
+>
+> In short, for errors that are a foreseeable and ordinary part of the
+> execution of your production code and that are not omnipresent,
+> error-aware return types are typically the right solution.
+
+In the case of Mirage, it is typically the case that every operation
+may fail and that, since the interfaces are abstract, we cannot know
+all the ways they may fail. This seems like a good argument for using
+exceptions.
+
+However, we still want to use explicit variants for expected
+(non-exceptional) cases that callers will probably want to handle.
+For example,
+[KV_RO](https://mirage.github.io/mirage-kv/Mirage_kv.html#TYPEerror)'s
+`Unknown_key` case is something code will often want to use, e.g. to
+provide a default value. This shouldn't even be considered an error.
+
+### Errors and Results
+
+Since 4.0.3, the OCaml standard library has introduced the
+[result
+type](https://caml.inria.fr/pub/docs/manual-ocaml/libref/Pervasives.html#TYPEresult):
+
+```ocaml
+type ('a, 'b) result = Ok of 'a | Error of 'b
+```
+
+A few useful libraries have also been released with useful
+combinators on result values, including Daniel Bünzli's
+[Rresult](http://erratique.ch/software/rresult/doc/Rresult.html) and
```

Comment:
Could you please link on http://erratique.ch/software/rresult/doc this will be more `odoc` proof.